lms/djangoapps/verify_student/models.py

```diff
@@ -1112,8 +1112,13 @@ def get_or_create_verification_checkpoint(cls, course_id, checkpoint_location):
         Returns:
             VerificationCheckpoint object if exists otherwise None
         """
-        checkpoint, __ = cls.objects.get_or_create(course_id=course_id, checkpoint_location=checkpoint_location)
-        return checkpoint
+        try:
+            with transaction.atomic():
+                checkpoint, __ = cls.objects.get_or_create(course_id=course_id, checkpoint_location=checkpoint_location)
+                return checkpoint
+        except IntegrityError:
+            # Record already exists - just return it.
+            return cls.objects.get(course_id=course_id, checkpoint_location=checkpoint_location)
```

Comment:
Since this method is called from an XBlock service, I doubt we can use `outer_atomic` here. So if an `IntegrityError` is raised we cannot do anything about it. The only thing I can think of doing here is using `READ COMMITTED` for the view which is calling this method. `courseware.views.index` is already using `READ COMMITTED` so that would make an `IntegrityError` here very unlikely. If it does happen I would let it propogate upwards for now.